File: web/meta-observability.js
const metaObsState={publications:new Map(),paid:new Map(),busy:new Set(),projectId:null};

function metaObsProjectId(){return state.current?.project?.id||null}
function metaObsPublished(){return (state.current?.publications||[]).filter(row=>row.status==='PUBLISHED'&&row.remote_id)}
function metaObsPaid(){return (state.current?.paid_media||[]).filter(row=>row.campaign_id||row.adset_id||row.creative_id||row.ad_id)}
function metaObsText(value,fallback='—'){return value===null||value===undefined||value===''?fallback:String(value)}
function metaObsNumber(value){const number=Number(value);return Number.isFinite(number)?number.toLocaleString():metaObsText(value)}

function ensureMetaObservabilityPanel(){
  const distribution=$('#social-distribution');if(!distribution)return null;
  let panel=$('#meta-observability');if(panel)return panel;
  panel=document.createElement('section');panel.id='meta-observability';panel.className='composer-card meta-observability-card';
  panel.innerHTML=`
    <div class="composer-head meta-observability-head">
      <div><p class="eyebrow">WAVE 24 · OBSERVABILIDAD META</p><h4>Verifica estado remoto y métricas sin modificar campañas</h4></div>
      <span id="meta-observability-safety" class="count-chip">READ ONLY</span>
    </div>
    <div class="meta-observability-toolbar">
      <p class="muted">Sólo hace lecturas GET a Meta. No cambia estado, presupuesto, targeting, copy ni creativos.</p>
      <div class="toolbar">
        <label>Ventana Ads<select id="meta-observability-period"><option value="maximum">Histórico disponible</option><option value="last_30d">Últimos 30 días</option><option value="last_14d">Últimos 14 días</option><option value="last_7d">Últimos 7 días</option><option value="today">Hoy</option><option value="yesterday">Ayer</option></select></label>
        <button id="meta-observability-refresh" type="button">Verificar recientes</button>
      </div>
    </div>
    <div class="grid-two meta-observability-grid">
      <div>
        <div class="composer-row-head"><strong>Publicaciones confirmadas</strong><span id="meta-observability-publication-count">0</span></div>
        <div id="meta-observability-publications" class="results"></div>
      </div>
      <div>
        <div class="composer-row-head"><strong>Pauta remota</strong><span id="meta-observability-paid-count">0</span></div>
        <div id="meta-observability-paid" class="results"></div>
      </div>
    </div>
    <p id="meta-observability-note" class="microcopy">Instagram muestra insights orgánicos disponibles; Facebook Reel verifica el estado oficial de procesamiento/publicación; Ads muestra estado remoto e Insights del Ad.</p>`;
  const paidList=distribution.querySelector('.paid-media-list-card');
  if(paidList)paidList.insertAdjacentElement('afterend',panel);else distribution.append(panel);
  $('#meta-observability-refresh').addEventListener('click',verifyRecentMetaObservability);
  $('#meta-observability-period').addEventListener('change',()=>{metaObsState.paid.clear();renderMetaObservability()});
  return panel;
}

function metaObsBusyKey(kind,id){return `${kind}:${id}`}
function metaObsButton(label,handler,disabled=false){const button=el('button','',label);button.type='button';button.disabled=disabled;button.addEventListener('click',handler);return button}
function metaObsMetricChip(label,value){const chip=el('span','meta-observability-metric');chip.append(el('strong','',metaObsText(value,'0')),el('small','',label));return chip}

function metaObsPublicationDetail(row,result){
  if(!result)return 'Sin verificar en esta sesión.';
  if(!result.available)return result.reason||'Meta todavía no tiene un objeto remoto verificable.';
  if(row.channel==='facebook_page'&&row.kind==='reel')return `Estado Reel: ${metaObsText(result.remote_state)}`;
  const remote=result.remote||{};return [result.remote_state,remote.media_product_type||remote.media_type,remote.timestamp||remote.created_time].filter(Boolean).join(' · ')||'Objeto remoto presente';
}

function renderMetaObsPublication(row,root){
  const result=metaObsState.publications.get(row.id),busy=metaObsState.busy.has(metaObsBusyKey('publication',row.id));
  const item=el('div','result-item meta-observability-item');
  const head=el('div','meta-observability-item-head');
  const copy=el('div','');copy.append(el('strong','',`${row.channel==='instagram'?'Instagram':'Facebook'} · ${row.kind}`),el('p','',`Meta ${row.remote_id}`));
  const button=metaObsButton(busy?'Verificando…':'Verificar',()=>verifyMetaPublication(row.id),busy||!socialState.status?.configured);head.append(copy,button);item.append(head,el('p','muted',metaObsPublicationDetail(row,result)));
  const insights=result?.insights||{};const metricNames=['reach','views','likes','comments','shares','saved','total_interactions'];const present=metricNames.filter(name=>insights[name]!==undefined);
  if(present.length){const metrics=el('div','meta-observability-metrics');for(const name of present)metrics.append(metaObsMetricChip(name.replaceAll('_',' '),metaObsNumber(insights[name])));item.append(metrics)}
  if(result?.metric_errors&&Object.keys(result.metric_errors).length)item.append(el('p','microcopy',`${Object.keys(result.metric_errors).length} métrica(s) no disponibles para este tipo de medio.`));
  root.append(item);
}

function metaObsObjectLine(label,row){if(!row)return `${label}: sin ID remoto`;return `${label}: ${metaObsText(row.observed_state||row.status||row.effective_status)} · ${metaObsText(row.id)}`}
function renderMetaObsPaid(row,root){
  const result=metaObsState.paid.get(row.id),busy=metaObsState.busy.has(metaObsBusyKey('paid',row.id));
  const item=el('div','result-item meta-observability-item');const head=el('div','meta-observability-item-head');
  const copy=el('div','');copy.append(el('strong','',row.campaign_name||'Pauta Meta'),el('p','',`${row.status} · ${row.ad_id?`Ad ${row.ad_id}`:'estructura parcial'}`));
  head.append(copy,metaObsButton(busy?'Verificando…':'Estado + métricas',()=>verifyMetaPaid(row.id),busy||!socialState.status?.configured));item.append(head);
  if(!result){item.append(el('p','muted','Sin verificar en esta sesión.'));root.append(item);return}
  if(!result.available){item.append(el('p','muted',result.reason||'Sin objetos remotos.'));root.append(item);return}
  const safety=result.safety||{};const safetyText=safety.explicit_active_detected?'ALERTA: Meta reportó un objeto configurado ACTIVE.':safety.configured_paused===true?'Campaign, Ad Set y Ad confirmados PAUSED.':'Estado remoto leído; revisa la jerarquía.';
  item.append(el('p',safety.explicit_active_detected?'meta-observability-alert':'muted',safetyText));
  const objects=result.objects||{};const lines=el('div','meta-observability-object-lines');lines.append(el('span','',metaObsObjectLine('Campaign',objects.campaign)),el('span','',metaObsObjectLine('Ad Set',objects.adset)),el('span','',metaObsObjectLine('Creative',objects.creative)),el('span','',metaObsObjectLine('Ad',objects.ad)));item.append(lines);
  const insights=result.insights||{};const metrics=el('div','meta-observability-metrics');for(const [label,key] of [['Impresiones','impressions'],['Alcance','reach'],['Clics','clicks'],['Gasto','spend'],['CTR','ctr'],['CPM','cpm']]){if(insights[key]!==undefined)metrics.append(metaObsMetricChip(label,metaObsNumber(insights[key])))}if(metrics.childElementCount)item.append(metrics);
  root.append(item);
}

function renderMetaObservability(){
  const panel=ensureMetaObservabilityPanel();if(!panel||!state.current?.project)return;
  const projectId=metaObsProjectId();if(metaObsState.projectId!==projectId){metaObsState.projectId=projectId;metaObsState.publications.clear();metaObsState.paid.clear();metaObsState.busy.clear()}
  const publications=metaObsPublished(),paid=metaObsPaid();$('#meta-observability-publication-count').textContent=String(publications.length);$('#meta-observability-paid-count').textContent=String(paid.length);
  const publicationRoot=$('#meta-observability-publications');publicationRoot.replaceChildren();for(const row of [...publications].reverse())renderMetaObsPublication(row,publicationRoot);if(!publications.length)publicationRoot.append(el('p','muted','Todavía no hay publicaciones PUBLISHED con remote_id.'));
  const paidRoot=$('#meta-observability-paid');paidRoot.replaceChildren();for(const row of [...paid].reverse())renderMetaObsPaid(row,paidRoot);if(!paid.length)paidRoot.append(el('p','muted','Todavía no hay estructura de pauta remota.'));
  const paidResults=[...metaObsState.paid.values()];const active=paidResults.some(result=>result?.safety?.explicit_active_detected);const allPaused=paidResults.length>0&&paidResults.every(result=>result?.safety?.configured_paused===true&&!result?.safety?.explicit_active_detected);const badge=$('#meta-observability-safety');badge.textContent=active?'ALERTA ACTIVE':allPaused?'PAUSED CONFIRMADO':'READ ONLY';badge.classList.toggle('ok',allPaused);badge.classList.toggle('meta-observability-danger',active);
}

async function verifyMetaPublication(publicationId){
  const projectId=metaObsProjectId();if(!projectId)return;const key=metaObsBusyKey('publication',publicationId);metaObsState.busy.add(key);renderMetaObservability();
  try{const result=await api(`/api/projects/${encodeURIComponent(projectId)}/publications/${encodeURIComponent(publicationId)}/observability`);metaObsState.publications.set(publicationId,result);toast('Estado de publicación verificado en Meta')}catch(err){toast(`Meta: ${err.message}`)}finally{metaObsState.busy.delete(key);renderMetaObservability()}
}
async function verifyMetaPaid(draftId){
  const projectId=metaObsProjectId();if(!projectId)return;const key=metaObsBusyKey('paid',draftId);metaObsState.busy.add(key);renderMetaObservability();
  try{const preset=$('#meta-observability-period')?.value||'maximum';const result=await api(`/api/projects/${encodeURIComponent(projectId)}/paid-media/${encodeURIComponent(draftId)}/observability?date_preset=${encodeURIComponent(preset)}`);metaObsState.paid.set(draftId,result);if(result?.safety?.explicit_active_detected)toast('ALERTA: Meta reportó un objeto ACTIVE');else toast('Estado de pauta verificado en Meta')}catch(err){toast(`Meta: ${err.message}`)}finally{metaObsState.busy.delete(key);renderMetaObservability()}
}
async function verifyRecentMetaObservability(){
  if(!socialState.status?.configured){toast('Conecta Meta antes de verificar estado remoto');return}
  const tasks=[...metaObsPublished().slice(-5).map(row=>['publication',row.id]),...metaObsPaid().slice(-5).map(row=>['paid',row.id])];if(!tasks.length){toast('No hay objetos remotos para verificar');return}
  for(const [kind,id] of tasks){if(kind==='publication')await verifyMetaPublication(id);else await verifyMetaPaid(id)}
}

function metaObsWatch(){const title=$('#active-project-name');if(title)new MutationObserver(()=>setTimeout(renderMetaObservability,0)).observe(title,{childList:true,characterData:true,subtree:true});setInterval(()=>{if(metaObsProjectId())renderMetaObservability()},3000);setTimeout(renderMetaObservability,0)}
if(document.readyState==='loading')document.addEventListener('DOMContentLoaded',metaObsWatch,{once:true});else metaObsWatch();
globalThis.renderMetaObservability=renderMetaObservability;

File: src/binario_marketing/meta_observability.py
from __future__ import annotations

from dataclasses import asdict
from typing import Any

from .meta_graph import MetaGraphClient, MetaGraphError
from .paid_media_store import PaidMediaDraft
from .social_store import Publication


_INSTAGRAM_MEDIA_METRICS = (
    "reach",
    "views",
    "likes",
    "comments",
    "shares",
    "saved",
    "total_interactions",
)
_AD_INSIGHT_FIELDS = (
    "ad_id",
    "date_start",
    "date_stop",
    "impressions",
    "reach",
    "clicks",
    "spend",
    "cpc",
    "cpm",
    "ctr",
    "frequency",
    "outbound_clicks",
    "actions",
    "video_play_actions",
)
_DATE_PRESETS = {
    "today",
    "yesterday",
    "last_7d",
    "last_14d",
    "last_30d",
    "this_month",
    "last_month",
    "maximum",
}


def _insight_value(row: dict[str, Any]) -> Any:
    values = row.get("values")
    if isinstance(values, list) and values:
        tail = values[-1]
        if isinstance(tail, dict) and "value" in tail:
            return tail["value"]
    return row.get("value")


def _normalize_insights(payload: dict[str, Any]) -> dict[str, Any]:
    rows = payload.get("data", [])
    if not isinstance(rows, list):
        raise MetaGraphError("Meta returned invalid insights data")
    result: dict[str, Any] = {}
    for row in rows:
        if not isinstance(row, dict):
            continue
        name = str(row.get("name") or "").strip()
        if name:
            result[name] = _insight_value(row)
    return result


def _remote_state(payload: dict[str, Any]) -> str:
    for key in ("configured_status", "status", "effective_status"):
        value = str(payload.get(key) or "").strip().upper()
        if value:
            return value
    return "UNKNOWN"


def _is_explicitly_active(payload: dict[str, Any]) -> bool:
    return any(str(payload.get(key) or "").strip().upper() == "ACTIVE" for key in ("configured_status", "status"))


def _is_configured_paused(payload: dict[str, Any]) -> bool | None:
    values = [str(payload.get(key) or "").strip().upper() for key in ("configured_status", "status")]
    values = [value for value in values if value]
    if not values:
        return None
    return all(value == "PAUSED" for value in values)


class MetaObservability:
    """Read-only Meta state and insight readback. Never mutates provider objects."""

    def __init__(self, client: MetaGraphClient):
        self.client = client

    @classmethod
    def from_env(cls) -> "MetaObservability":
        return cls(MetaGraphClient.from_env())

    def _instagram_media_insights(self, media_id: str, instagram_id: str) -> tuple[dict[str, Any], dict[str, str]]:
        token = self.client._instagram_token(instagram_id)
        metrics = ",".join(_INSTAGRAM_MEDIA_METRICS)
        try:
            payload = self.client._request("GET", f"{media_id}/insights", {"metric": metrics}, token=token)
            return _normalize_insights(payload), {}
        except MetaGraphError:
            # Meta can reject a mixed metric set when a metric is not valid for a
            # particular media type. Retry independently so supported metrics survive.
            values: dict[str, Any] = {}
            errors: dict[str, str] = {}
            for metric in _INSTAGRAM_MEDIA_METRICS:
                try:
                    payload = self.client._request("GET", f"{media_id}/insights", {"metric": metric}, token=token)
                    values.update(_normalize_insights(payload))
                except MetaGraphError as exc:
                    errors[metric] = str(exc)
            return values, errors

    def publication(self, row: Publication) -> dict[str, Any]:
        base = {
            "publication_id": row.id,
            "channel": row.channel,
            "kind": row.kind,
            "local_status": row.status,
            "remote_id": row.remote_id,
            "available": False,
            "remote": None,
            "insights": {},
            "metric_errors": {},
        }
        if row.status != "PUBLISHED" or not row.remote_id:
            base["reason"] = "publication has no confirmed remote Meta object yet"
            return base

        if row.channel == "facebook_page":
            token = self.client._page_token(row.target_id)
            if row.kind == "reel":
                remote = self.client._request("GET", row.remote_id, {"fields": "status"}, token=token)
                base.update({"available": True, "remote": remote, "remote_state": str((remote.get("status") or {}).get("video_status") or "UNKNOWN").upper()})
                return base
            try:
                remote = self.client._request("GET", row.remote_id, {"fields": "id,created_time,permalink_url"}, token=token)
            except MetaGraphError:
                remote = self.client._request("GET", row.remote_id, {"fields": "id"}, token=token)
            base.update({"available": True, "remote": remote, "remote_state": "PRESENT"})
            return base

        if row.channel == "instagram":
            token = self.client._instagram_token(row.target_id)
            remote = self.client._request(
                "GET",
                row.remote_id,
                {"fields": "id,media_type,media_product_type,permalink,timestamp,caption"},
                token=token,
            )
            insights, metric_errors = self._instagram_media_insights(row.remote_id, row.target_id)
            base.update({
                "available": True,
                "remote": remote,
                "remote_state": "PRESENT",
                "insights": insights,
                "metric_errors": metric_errors,
            })
            return base

        base["reason"] = "unsupported publication channel"
        return base

    def paid_media(self, row: PaidMediaDraft, *, date_preset: str = "maximum") -> dict[str, Any]:
        preset = str(date_preset or "maximum").strip().lower()
        if preset not in _DATE_PRESETS:
            raise ValueError("unsupported Meta insights date_preset")

        payload: dict[str, Any] = {
            "draft_id": row.id,
            "local_status": row.status,
            "date_preset": preset,
            "available": False,
            "objects": {},
            "insights": {},
            "safety": {
                "activation_endpoint_present": False,
                "explicit_active_detected": False,
                "configured_paused": None,
            },
        }
        remote_ids = {
            "campaign": row.campaign_id,
            "adset": row.adset_id,
            "creative": row.creative_id,
            "ad": row.ad_id,
        }
        if not any(remote_ids.values()):
            payload["reason"] = "paid-media draft has no remote Meta objects yet"
            return payload

        fields = {
            "campaign": "id,name,objective,configured_status,effective_status,status,created_time,updated_time",
            "adset": "id,name,campaign_id,configured_status,effective_status,status,daily_budget,created_time,updated_time",
            "creative": "id,name",
            "ad": "id,name,campaign_id,adset_id,configured_status,effective_status,status,created_time,updated_time",
        }
        configured_paused: list[bool] = []
        explicit_active = False
        for kind, remote_id in remote_ids.items():
            if not remote_id:
                continue
            remote = self.client._request("GET", remote_id, {"fields": fields[kind]})
            remote["observed_state"] = _remote_state(remote)
            payload["objects"][kind] = remote
            if kind in {"campaign", "adset", "ad"}:
                active = _is_explicitly_active(remote)
                explicit_active = explicit_active or active
                paused = _is_configured_paused(remote)
                if paused is not None:
                    configured_paused.append(paused)

        if row.ad_id:
            insights_payload = self.client._request(
                "GET",
                f"{row.ad_id}/insights",
                {"fields": ",".join(_AD_INSIGHT_FIELDS), "date_preset": preset},
            )
            rows = insights_payload.get("data", [])
            if not isinstance(rows, list):
                raise MetaGraphError("Meta returned invalid paid-media insights data")
            payload["insights"] = rows[0] if rows and isinstance(rows[0], dict) else {}

        payload["available"] = True
        payload["safety"] = {
            "activation_endpoint_present": False,
            "explicit_active_detected": explicit_active,
            "configured_paused": (all(configured_paused) if configured_paused else None),
        }
        payload["local"] = {
            "id": row.id,
            "status": row.status,
            "campaign_id": row.campaign_id,
            "adset_id": row.adset_id,
            "creative_id": row.creative_id,
            "ad_id": row.ad_id,
        }
        return payload


__all__ = ["MetaObservability"]
